packages/block-library/src/template-part/index.php

```diff
@@ -56,12 +65,36 @@ function render_block_core_template_part( $attributes ) {
 		}
 	}
 
-	if ( is_null( $content ) ) {
-		return 'Template Part Not Found';
+	if ( ! $template_part_id || is_null( $content ) ) {
+		return __( 'Template Part not found.' );
+	}
+
+	if ( isset( $seen_ids[ $template_part_id ] ) ) {
+		if ( ! is_admin() ) {
+			trigger_error(
+				sprintf(
+					// translators: %s are the block attributes.
+					__( 'Could not render Template Part block with the attributes: <code>%s</code>. Block cannot be rendered inside itself.' ),
+					wp_json_encode( $attributes )
+				),
+				E_USER_WARNING
+			);
+		}
+
+		// WP_DEBUG_DISPLAY must only be honored when WP_DEBUG. This precedent
+		// is set in `wp_debug_mode()`.
+		$is_debug = defined( 'WP_DEBUG' ) && WP_DEBUG &&
+			defined( 'WP_DEBUG_DISPLAY' ) && WP_DEBUG_DISPLAY;
+		return $is_debug ?
+			// translators: Visible only in the front end, this warning takes the place of a faulty block.
+			__( '[block rendering halted]' ) :
+			'';
 	}
 
 	// Run through the actions that are typically taken on the_content.
-	$content = do_blocks( $content );
+	$seen_ids[ $template_part_id ] = true;
+	$content                       = do_blocks( $content );
+	unset( $seen_ids[ $template_part_id ] );
```

Comment:
I'm a bit confused with this. Why do you unset after setting?

Reply:
You need to unset template part ids once the child blocks were processed, so other blocks in the same level (siblings) would render. It's the case like this pseudocode:

```jsx
<>
    <TemplatePart id="A" />
    <TemplatePart id="A" />
</>
```

If you wouldn't unset, then it would consider the second occurrence as already seen and it would prevent its rendering.

Reply:
We unset after calling `do_blocks`. :) So the flow becomes:

1. Set _seen_ to prevent future loops
2. "Do" the content, i.e. traverse and render any inner blocks
3. Unset _seen_, so that we don't prevent legitimate use cases, like the same block being rendered in distinct stacks. Imagine the following:

* Document
  * Header
    * **Site Title + Logo** template part
  * Main
    * Post content
  * Footer
    * **Site Title + Logo** template part

Reply:
Thanks both!

Reply:
(GitHub didn't properly reload, so I didn't realise Grzegorz had already replied 🤦)